File: src/driver/generic/PIDController.cpp
#include "rclcpp/rclcpp.hpp"
#include "PIDController.hpp"

PID::PID(float i_kp, float i_ki, float i_kd)
{
    kp = i_kp;
    ki = i_ki;
    kd = i_kd;
    error = 0.0f;
    prevErr = 0.0f;
    clock(RCL_ROS_TIME);
    prevTime = clock.now();
    result = 0.0f;
}

float PID::step(float offset)
{
    float P = 0.0f;
    float I = 0.0f;
    float D = 0.0f;

    rclcpp::Time currTime = clock.now();
    rclcpp::Duration dt = currTime - prevTime;

    if(dt < MIN_TIME)
    {
        dt = MIN_TIME;
    } 
    else if(dt > MAX_TIME)
    {
        dt = MAX_TIME;
    }

    error = offset;
    P = error*kp;

    I += ki*error*dt;

    if(I < MIN_I)
    {
        I = MIN_I;
    }
    else if(I > MAX_I)
    {
        I = MAX_I;
    }

    D = (error-prevErr)/dt;

    result = P + I + D;
    prevErr = error;
    prevTime = currTime;
    return result;
}

void PID::resetErr()
{
    prevErr = 0.0f;
}

File: src/kestrel_control/python/kestrel_control/pid_controller.py
import rclpy
from rclpy.clock import Clock, ClockType
from rclpy.duration import Duration

class PIDController():
    #Clamping variables for frame rate (so it is not 0)
    MAX_TIME = Duration(nanoseconds = 100_000_000)
    MIN_TIME = Duration(nanoseconds = 1_000_000)

    #Clamping variables for integral
    MAX_I = 1.0
    MIN_I = -1.0

    def __init__ (self, kp, ki, kd):
        self.kp = kp
        self.ki = ki
        self.kd = kd

        self.clock = Clock(clock_type=ClockType.SYSTEM_TIME)
        self.prev_time = self.clock.now()

        self.prev_error = 0.0

        self.i = 0.0

    def compute_command(self, offset): #or whatever name
        error = offset #error test, could be distance or motor offset

        curr_time = self.clock.now() # if dt needs to be taken out just delete this section and add to func def
        dt = curr_time-self.prev_time

        #Clamping duration bc max min do not understand ROS2
        if dt > self.MAX_TIME :
            dt = self.MAX_TIME
        elif dt < self.MIN_TIME :
            dt = self.MIN_TIME

        dt = dt.nanoseconds / 1e9

        p = self.kp * error #Calculate correction

        self.i += self.ki * error * dt #Handle over time error
        self.i = max(min(self.i, self.MAX_I), self.MIN_I) #Clamp I so not overshooting, picking out smallest num then biggest num
        
        derivative = (error - self.prev_error)/dt #Find derivative for d part
        d = self.kd * derivative #Handle overshoot

        pid = p + self.i + d #Put it all together

        self.prev_time = curr_time #Update prev_time
        self.prev_error = error #Set new previous error

        return pid
    
    #Just in case error should be reset
    def reset_error(self) :
        self.prev_error = 0.0
